samples: subsys: input: draw_touch_events: fix overflow in clear_screen loop

CharlesDias opened the pull request:
Problem: On STM32U5G9J-DK1 + st_lcd_dsi_mb1835, the sample could BUS FAULT with SCREEN_WIDTH_TO_CROSS_DIM=25 and show only a horizontal line with 20.

~~Root cause: The tile buffer BPP is derived from the devicetree pixel format (e.g., RGB888 = 3 BPP), but the driver may be using a different current pixel format.~~

~~Fix: Explicitly set the driver pixel format to the DT-selected format at init.~~

**EDIT:**
The `clear_screen` loop used fixed-size tiles (CROSS_DIM). When WIDTH/HEIGHT aren’t multiples of CROSS_DIM, the right/bottom tiles exceeded the panel bounds, causing the driver’s row copy (memcpy) to write past the framebuffer and fault.

Fix: clip only the edge tiles

- width = min(CROSS_DIM, WIDTH - x)
- height = min(CROSS_DIM, HEIGHT - y)
- pitch remains CROSS_DIM (source stride)
- buf_size = width * height * BPP

Prevents out-of-bounds writes on edge tiles.

Related with:
- https://github.com/zephyrproject-rtos/zephyr/pull/98717

https://github.com/user-attachments/assets/993ed0d1-ee71-4d50-bfe4-0e9eb88f03c9

Fixes #98922

CharlesDias committed:
samples: input: draw_touch_events: avoid avoid OOB writes

When CROSS_DIM doesn’t evenly divide the panel WIDTH/HEIGHT, the last
tile on the right/bottom edge can extend past the display bounds.

Fix by clipping the edge tiles.

Signed-off-by: Charles Dias <charlesdias.cd@outlook.com>

cfriedt commented:
@CharlesDias - can you link a bug report to this issue? It's required for all bug fixes post RC2.

CharlesDias commented:
Hi @cfriedt. 

I opened the bug:
- https://github.com/zephyrproject-rtos/zephyr/issues/98922

CharlesDias commented:
Hi, @fabiobaltieri, @faxe1008, and @cfriedt. 

I've changed this fix to DRAFT because, although this modification seems to have resolved the problem, I suspect the root cause may be something else. I am currently investigating further. I apologize for the inconvenience!

CharlesDias committed:
boards: shields: st_lcd_dsi_mb1835: set LTDC pixel-format

Set zephyr_lcd_controller pixel-format to PANEL_PIXEL_FORMAT_ARGB_8888
to match the driver configuration.

Signed-off-by: Charles Dias <charlesdias.cd@outlook.com>

CharlesDias commented:
After further investigation, I found that the BUS FAULT was caused by the clear_screen loop writing past the framebuffer. When CROSS_DIM does not evenly divide the panel WIDTH/HEIGHT, the last tile on the right/bottom extends beyond the framebuffer row/column.

Backtrace

```
#12 0x0800a3ea in memcpy ()
#13 0x08007ece in stm32_ltdc_write (dev=<optimized out>, x=<optimized out>, y=<optimized out>, desc=0x20000000 <buf_desc>, buf=0x800aff4 <buffer_cross_empty>) at /home/charlesdias/zephyrproject/zephyr/drivers/display/display_stm32_ltdc.c:233
#14 0x0800056e in display_write (buf=0x800aff4 <buffer_cross_empty>, desc=0x20000000 <buf_desc>, y=190, x=399, dev=<optimized out>) at /home/charlesdias/zephyrproject/zephyr/include/zephyr/drivers/display.h:288
#15 clear_screen () at /home/charlesdias/zephyrproject/zephyr/samples/subsys/input/draw_touch_events/src/main.c:93
#16 0x08000694 in main () at /home/charlesdias/zephyrproject/zephyr/samples/subsys/input/draw_touch_events/src/main.c:156
```


**Solution**:
For each x,y tile, it now creates a local descriptor with:

- width = min(CROSS_DIM, WIDTH - x)
- height = min(CROSS_DIM, HEIGHT - y)
- pitch remains CROSS_DIM (source stride in pixels)
- buf_size = width * height * BPP

This guarantees we never copy past the framebuffer boundaries for tiles at the right/bottom edges.

**Additional fix (pixel format)**
Additionally, the zephyr_lcd_controller pixel-format did not match the driver configuration, which expects PANEL_PIXEL_FORMAT_ARGB_8888.

**Solution (pixel format)**
Set the zephyr_lcd_controller pixel-format to PANEL_PIXEL_FORMAT_ARGB_8888 to match the driver configuration.

avolmat-st commented:
> > I'm going to open a new PR in order to add display API return value check as well as protect the LTDC driver memcpy.
> 
> Hi @avolmat-st. Thanks for your inputs!
> 
> If needed, I can test your new PR on my STM32U5G9J-DK1 board. Please let me know. Thank you!

Thanks @CharlesDias, I've just raised the PR #99166 which tackles this point. Could you recheck your side with your setup, and also for the test purpose only revert your fix commit in the sample in order to ensure that the LTDC driver is properly protecting this case ?
